app/models/box_form.rb

```diff
@@ -0,0 +1,101 @@
+class BoxForm
+  attr_reader :box, :batch_numbers, :batch_errors
+
+  delegate :purpose, :purpose=, to: :box
+  delegate :model_name, :errors, to: :box
+
+  def self.build(navigation_context, params = nil)
+    box = Box.new(
+      institution: navigation_context.institution,
+      site: navigation_context.site,
+    )
+
+    if params
+      batch_numbers = params.delete(:batch_numbers) || {}
+      box.attributes = params
+    else
+      batch_numbers = {}
+    end
+
+    new(box, batch_numbers)
+  end
+
+  def initialize(box, batch_numbers)
+    @box = box
+    @batch_numbers = batch_numbers
+    @batch_errors = {}
+  end
+
+  def batches=(relation)
+    @batches = {}
+    records = relation.to_a
+
+    @batch_numbers.each do |key, batch_number|
+      if batch = records.find { |b| b.batch_number == batch_number }
+        @batches[key] = batch
+      elsif batch_number.present?
+        @batch_errors[key] = "batch doesn't exist"
+      end
+    end
+  end
+
+  def build_samples
+    case @box.purpose
+    when "LOD"
+      if batch = get_batch("0") { "please select a batch" }
+        @box.build_samples(batch, exponents: 1..8, replicas: 3)
+      end
+
+    when "Variants"
+      @batches.each_value do |batch|
+        @box.build_samples(batch, exponents: [1, 4, 8], replicas: 3)
+      end
+
+    when "Challenge"
+      if batch = get_batch("0") { "please select a virus batch" }
+        @box.build_samples(batch, exponents: [1, 4, 8], replicas: 18)
+      end
+      @batches.each do |key, batch|
+        @box.build_samples(batch, exponents: [1, 4, 8], replicas: 3) unless key == "0"
+      end
```

Comment:
I'm a bit sceptical of just numbering the samples and giving the number zero a special meaning instead of using an explicit name (`virus`). But changing this would probably complicate the logic for loading batches.
Just a thought for consideration.

Reply:
I have the same uneasiness. I think I still prefer a flat list of batch numbers, since it's more generic, and we can interpret it in any way internally (it could even be customized by deployment).

Yet, now that I have a BoxForm, I could have a `virus_batch_number` or `specific_batch_number`.

I'd love to see the list of 3 new purposes to decide further.

Reply:
I named the keys instead of relying on number indexes. It doesn't change anything, because we already expected a `Hash<String, String>`, but it improves the overall readability!